README.md
# Simple (HTTP) Proxy

Simple HTTP proxy for node.js that pipelines content from back-end to front-end server with minimal buffer and minimal latency.

## Installation

```bash
npm install
npm start
```

## Configuration

| Name              | Type   | Default                      | Description                                                                                                                                     |
| ----------------- | ------ | ---------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------- |
| **HOSTNAME**      | `str`  | `127.0.0.1`                  | The address of the server that serves the app.                                                                                                  |
| **PORT**          | `int`  | `3000`                       | The port the server will listen at.                                                                                                             |
| **KEY**           | `str`  | `null`                       | Secret key that should be passed in protected calls so that the server side "trusts" the client side (authentication).                          |
| **TARGET**        | `str`  | `https://app.platforme.com/` | The target server that the server will proxy to.                                                                                                |
| **PRX_TARGET**    | `str`  | `https://app.platforme.com/` | The same as `TARGET`.                                                                                                                           |
| **PRX_KEEPALIVE** | `int`  | `600000`                     | The number of milliseconds to keep the back-end connections alive in the connection pool                                                        |
| **WHITELIST**     | `list` | `[]`                         | A list of query params and allowed values. Query params not in the list are not checked. (e.g. `WHITELIST="param1:val1,val2;param2:val1,val2"`) |

## License

Simple (HTTP) Proxy is currently licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/).

## Build Automation

[![Build Status](https://github.com/ripe-tech/simple-proxy/workflows/Main%20Workflow/badge.svg)](https://github.com/ripe-tech/simple-proxy/actions)
[![License](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://www.apache.org/licenses/)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

*

### Changed

* Removed node 10 and 11 and added node 18

### Fixed

* Fix eslint dependencies problems

## [0.2.4] - 2022-03-31

### Fixed

* Issue related with empty body payloads and forwarding

## [0.2.3] - 2022-03-31

### Added

* New log information on Node.js version

## [0.2.2] - 2022-03-31

### Fixed

* Better `body` attribute validation

## [0.2.1] - 2022-03-30

### Changed

* Minimal code tune

## [0.2.0] - 2022-03-30

### Added

* Added a query param whitelisting mechanism for incoming requests (`WHITELIST`)
* Passing request body in the requests

### Changed

* Bumped packages
* De-bumped eslint package, due to error


test/util/base.test.js
const assert = require("assert");

const config = require("../../lib/util/config");

const lib = require("../../lib");

describe("Base", function() {
    describe("#init()", function() {
        it("should do nothing", () => {
            lib.init();
        });
    });

    describe("#destroy", function() {
        it("should do nothing", () => {
            lib.destroy();
        });
    });

    describe("#verifyKey()", function() {
        beforeEach(async () => {
            await lib.start();
        });

        it("should allow a request if no server key is defined", () => {
            const key = config.conf.KEY;
            config.conf.KEY = null;
            try {
                lib.verifyKey({});
            } finally {
                config.conf.KEY = key;
            }
        });

        it("should allow a request if both the request query key and server keys match", () => {
            const key = config.conf.KEY;
            config.conf.KEY = "key";
            try {
                lib.verifyKey({
                    query: {
                        key: "key"
                    }
                });
            } finally {
                config.conf.KEY = key;
            }
        });

        it("should allow a request if both the request header key and server keys match", () => {
            const key = config.conf.KEY;
            config.conf.KEY = "key";
            try {
                lib.verifyKey({
                    query: {},
                    headers: {
                        "X-Proxy-Key": "key"
                    }
                });
            } finally {
                config.conf.KEY = key;
            }
        });

        it("should fail for a request with no defined key", () => {
            const key = config.conf.KEY;
            config.conf.KEY = "key";
            try {
                assert.rejects(
                    () =>
                        lib.verifyKey({
                            query: {},
                            headers: {}
                        }),
                    err => {
                        assert.strictEqual(err.name, "Error");
                        assert.strictEqual(err.message, "Invalid key");
                        return true;
                    }
                );
            } finally {
                config.conf.KEY = key;
            }
        });
    });

    describe("#verifyQuery", function() {
        beforeEach(async () => {
            await lib.start();
            config.conf.WHITELIST = {
                hello: ["world"],
                param3: ["1", "2"]
            };
        });

        it("should allow requests with no query params", () => {
            lib.verifyQuery({
                query: {}
            });
        });

        it("should allow requests with no white listed params", () => {
            lib.verifyQuery({
                query: {
                    param1: "value1",
                    param2: "value2"
                }
            });
        });

        it("should allow requests for white list compliant param values", () => {
            lib.verifyQuery({
                query: {
                    hello: "world"
                }
            });

            lib.verifyQuery({
                query: {
                    param3: "1"
                }
            });

            lib.verifyQuery({
                query: {
                    hello: "world",
                    param3: "2"
                }
            });
        });

        it("should fail for non white list compliant param values", async () => {
            await assert.rejects(
                async () =>
                    lib.verifyQuery({
                        query: {
                            hello: "unknown"
                        }
                    }),
                err => {
                    assert.strictEqual(err.name, "Error");
                    assert.strictEqual(err.message, "Restricted value 'unknown' for param 'hello'");
                    return true;
                }
            );
        });
    });
});


app.js
// requires the multiple libraries
const https = require("https");
const express = require("express");
const process = require("process");
const util = require("hive-js-util");
const info = require("./package");
const lib = require("./lib");
const request = require("request");

// builds the initial application object to be used
// by the application for serving
const app = express();

// creates the HTTP connection pool/agent that is going
// to be used by the proxy
const agent = new https.Agent({
    protocol: "https:",
    keepAlive: true,
    keepAliveMsecs: lib.conf.PRX_KEEPALIVE
});

process.on("SIGINT", function() {
    process.exit();
});

process.on("SIGTERM", function() {
    process.exit();
});

process.on("exit", () => {
    util.Logging.info("Exiting on user's request");
    lib.destroy();
});

app.use(
    express.raw({
        type: () => true,
        limit: "50mb"
    })
);

app.get("/info", (req, res, next) => {
    res.json({
        name: info.name,
        version: info.version,
        node: process.version
    });
});

app.options("*", (req, res, next) => {
    const headers = lib.proxyHeaders(req);
    Object.assign(headers, {
        "Content-Security-Policy":
            "default-src * ws://* wss://* data: blob:; script-src * 'unsafe-inline' 'unsafe-eval'; style-src * 'unsafe-inline';",
        "Access-Control-Allow-Headers": "*",
        "Access-Control-Allow-Methods": "GET, POST, PUT, DELETE, PATCH, HEAD, OPTIONS",
        "Access-Control-Allow-Origin": "*"
    });
    res.set(headers);
    res.send();
});

app.all("*", (req, res, next) => {
    async function clojure() {
        // makes sure that in case there's a secret key defined then
        // it must be properly set in the request
        lib.verifyKey(req);

        // makes sure that if there's GET query validation required
        // then it's ensured and exceptions are thrown in case of errors
        lib.verifyQuery(req);

        await new Promise((resolve, reject) => {
            try {
                // constructs the initial options object with the
                // processed headers and query string
                const options = {
                    agent: agent,
                    baseUrl: lib.conf.PRX_TARGET,
                    uri: req.path,
                    method: req.method,
                    headers: lib.proxyHeaders(req),
                    qs: req.query,
                    qsStringifyOptions: {
                        arrayFormat: "repeat"
                    },
                    forever: true
                };

                // in case there's a valid body defined for the request
                // then sets the body in the options
                if (req.body && Buffer.isBuffer(req.body) && req.body.length > 0) {
                    options.body = req.body;
                }

                // runs the changed request with the transformed values so
                // that they become compliant with the expected API
                request(options).on("finish", resolve).on("error", reject).pipe(res);
            } catch (err) {
                reject(err);
            }
        });
    }
    clojure().catch(next);
});

(async () => {
    await lib.start();
    try {
        app.listen(lib.conf.PORT, lib.conf.HOST, () => {
            try {
                util.Logging.info(`Running v${info.version} in Node.js ${process.version}`);
                util.Logging.info(`Listening on ${lib.conf.HOST}:${String(lib.conf.PORT)}`);
                lib.init();
            } catch (err) {
                util.Logging.error(err);
            }
        });
    } finally {
        await lib.stop();
    }
})();


lib/util/base.js
const config = require("./config");

const init = () => {};

const destroy = () => {};

/**
 * Verifies that the key present in the request matches
 * the one defined in the current configuration, ensuring
 * that proper security measures are in place.
 *
 * @param {any} req The request to retrieve the key.
 */
const verifyKey = req => {
    if (!config.conf.KEY) {
        return;
    }
    const _key = req.query.key || req.headers["X-Proxy-Key"] || null;
    if (config.conf.KEY === _key) {
        return;
    }
    throw new Error("Invalid key");
};

/**
 * Verifies that the request query params are white listed, ensuring
 * that proper security measures are in place.
 *
 * @param {any} req The request to retrieve the query params.
 */
const verifyQuery = req => {
    const WHITELIST = config.conf.WHITELIST;
    Object.keys(req.query).forEach(param => {
        const value = req.query[param];
        if (WHITELIST[param] !== undefined && !WHITELIST[param].includes(value)) {
            throw new Error(`Restricted value '${value}' for param '${param}'`);
        }
    });
};

module.exports = {
    init: init,
    destroy: destroy,
    verifyKey: verifyKey,
    verifyQuery: verifyQuery
};


lib/util/config.js
const util = require("hive-js-util");
const yonius = require("yonius");

const conf = {};

const start = async () => {
    await startBase();
    await Promise.all([startConfig(), startLogging()]);
};

const stop = async () => {};

const startBase = async () => {
    await yonius.load();
};

const startConfig = async () => {
    conf.HOST = yonius.conf("HOST", "127.0.0.1");
    conf.PORT = yonius.conf("PORT", 3000, "int");
    conf.KEY = yonius.conf("PROXY_KEY", null);
    conf.PRX_TARGET = yonius.conf("PROXY_TARGET", "https://app.platforme.com/");
    conf.PRX_TARGET = yonius.conf("PRX_TARGET", conf.PRX_TARGET);
    conf.PRX_KEEPALIVE = yonius.conf("PRX_KEEPALIVE", 600000, "int");
    conf.WHITELIST = _buildWhiteList();
};

const startLogging = async () => {
    const level = yonius.conf("LEVEL", "DEBUG").toUpperCase();

    const logger = util.Logging.getLogger(undefined, {
        level: util.Logging.constants[level]
    });

    if (util.Logging.ConsolaHandler.isReady()) {
        logger.addHandler(new util.Logging.ConsolaHandler());
        logger.setFormatter(new util.Logging.SimpleFormatter("{asctime} {message}"));
    } else {
        logger.addHandler(new util.Logging.StreamHandler());
        logger.setFormatter(new util.Logging.SimpleFormatter());
    }
};

const _buildWhiteList = () => {
    const whiteList = {};
    const WHITELIST = yonius.conf("WHITELIST", [], "list");
    WHITELIST.forEach(list => {
        const [param, values] = list.split(":", 2);
        whiteList[param] = values.split(",");
    });
    return whiteList;
};

module.exports = {
    conf: conf,
    start: start,
    stop: stop
};
